### ROSHAN BHATIA
7374093589

In [1]:
import pandas as pd
import numpy as np

#### IMPORTING XLSX FILE

In [2]:
df = pd.read_excel("uber_rides_data.xlsx")

#### Q1. What is the shape of given dataset?


In [3]:
df.shape

(200000, 8)

#### Q2. How many integer columns(by default) are given in the dataset?


In [4]:
column_data_types = df.dtypes

integer_columns = column_data_types[column_data_types == 'int64']

num_integer_columns = len(integer_columns)

print("Number of integer columns:", num_integer_columns)

Number of integer columns: 2


#### Q3. How many missing values exists in 'dropoff_longitude' column?


In [5]:
missing_values = df['dropoff_longitude'].isnull().sum()

print("Number of missing values in 'dropoff_longitude' column:", missing_values)

Number of missing values in 'dropoff_longitude' column: 1


# Q4. What is the data type of ' pickup_datetime' feature in your data?
 

In [6]:
pickup_datetime_dtype = df['pickup_datetime'].dtype

print("Data type of 'pickup_datetime' feature:", pickup_datetime_dtype)

Data type of 'pickup_datetime' feature: object


#### Q5. Which of the following is the correct syntax to convert 'pickup_datetime' to datetime datatype?

 Correct syntax to convert 'pickup_datetime' to datetime datatype :

In [7]:
# pd.to_datetime(df['pickup_datetime'])

#### Q6. Which function can be used to remove null values from the dataframe?

In [8]:
df_cleaned = df.dropna()

Cleaned dataframe:

In [9]:
df_cleaned.head()

ride_id  fare_amount          pickup_datetime  pickup_longitude  \
0  24238194          7.5  2015-05-07 19:52:06 UTC        -73.999817   
1  27835199          7.7  2009-07-17 20:04:56 UTC        -73.994355   
2  44984355         12.9  2009-08-24 21:45:00 UTC        -74.005043   
3  25894730          5.3  2009-06-26 08:22:21 UTC        -73.976124   
4  17610152         16.0  2014-08-28 17:47:00 UTC        -73.925023   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  
0        40.738354         -73.999512         40.723217                1  
1        40.728225         -73.994710         40.750325                1  
2        40.740770         -73.962565         40.772647                1  
3        40.790844         -73.965316         40.803349                3  
4        40.744085         -73.973082         40.761247                5

#### Q7. What is the average fare amount?


In [10]:
average_fare_amount = df_cleaned['fare_amount'].mean()

print("Average fare amount:", average_fare_amount)

Average fare amount: 11.359891549458371


#### Q8 .Calculate distance between each pickup and dropoff points using Haversine formula. 


#### What is the median haversine distance between pickup and dropoff location according to the given dataset?

In [11]:
#function to calculate Haversine distance
def haversine(lat1, lon1, lat2, lon2):
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])

    # Haversine formula
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = np.sin(dlat / 2) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2) ** 2
    c = 2 * np.arcsin(np.sqrt(a))
    r = 6371  # Radius of Earth in kilometers
    return c * r

In [12]:
df_cleaned['haversine_distance'] = df_cleaned.apply(lambda row: haversine(row['pickup_latitude'], row['pickup_longitude'],
                                                          row['dropoff_latitude'], row['dropoff_longitude']), axis=1)

median_haversine_distance = df_cleaned['haversine_distance'].median()

print("Median Haversine distance between pickup and dropoff locations:", median_haversine_distance, "kilometers")

Median Haversine distance between pickup and dropoff locations: 2.120992396182902 kilometers


<ipython-input-12-9694d3b031be>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['haversine_distance'] = df_cleaned.apply(lambda row: haversine(row['pickup_latitude'], row['pickup_longitude'],


#### Q9. What is the maximum haversine distance between pickup and dropoff location according to the given dataset?

In [13]:
max_haversine_distance =df_cleaned['haversine_distance'].max()

print("Maximum Haversine distance between pickup and dropoff locations:", max_haversine_distance, "kilometers")

Maximum Haversine distance between pickup and dropoff locations: 16409.239135313168 kilometers


#### Q10. How many rides have 0.0 haversine distance between pickup and dropoff location according to the given dataset?

In [14]:
zero_distance_rides =  df_cleaned[df_cleaned['haversine_distance'] == 0.0]
num_zero_distance_rides = len(zero_distance_rides)

print("Number of rides with 0.0 Haversine distance between pickup and dropoff locations:", num_zero_distance_rides)

Number of rides with 0.0 Haversine distance between pickup and dropoff locations: 5632


#### Q11. What is the mean 'fare_amount' for rides with 0 haversine distance?

In [15]:
zero_distance_rides =  df_cleaned[ df_cleaned['haversine_distance'] == 0.0]

mean_fare_amount_for_zero_distance_rides = zero_distance_rides['fare_amount'].mean()

print("Mean 'fare_amount' for rides with 0.0 Haversine distance:", mean_fare_amount_for_zero_distance_rides)

Mean 'fare_amount' for rides with 0.0 Haversine distance: 11.585317826704578


#### Do you sense something fishy? Try to analyze, and give your expert opinion in Jupyter Notebook.

Indeed, the difference in the mean fare amounts for rides with a Haversine distance of 0.0 (11.59) compared to the overall mean fare (11.35) does raise some suspicion. This difference suggests that there might be some inconsistency or anomaly in the data.

Here are a few possibilities to consider and further investigate:

1. Data Entry Errors: There could be data entry errors or inconsistencies in the dataset. Some rides might have been recorded with incorrect coordinates or fares.

2. Short Trips: Rides with 0.0 Haversine distance could represent very short trips, such as rides that start and end at the same location (e.g., within a parking lot or around a block). These trips might have a flat-rate or minimum fare, which could explain the higher mean fare for such trips.

3. Outliers: There may be outliers in the 'fare_amount' column for rides with 0.0 Haversine distance. A few exceptionally high fares for short trips could significantly affect the mean.

4. Data Preprocessing: It's possible that during data preprocessing or feature engineering, certain adjustments or transformations were applied to the fare data for these specific rides.

#### Q11. What is the maximum 'fare_amount' for a ride?


In [16]:
max_fare_amount = df_cleaned['fare_amount'].max()

print("Maximum 'fare_amount' for a ride:", max_fare_amount)

Maximum 'fare_amount' for a ride: 499.0


#### Q12. What is the haversine distance between pickup and dropoff location for the costliest ride?

In [17]:
costliest_ride = df[df['fare_amount'] == df_cleaned['fare_amount'].max()]

# Extract the latitude and longitude coordinates for pickup and dropoff
pickup_latitude = costliest_ride['pickup_latitude'].values[0]
pickup_longitude = costliest_ride['pickup_longitude'].values[0]
dropoff_latitude = costliest_ride['dropoff_latitude'].values[0]
dropoff_longitude = costliest_ride['dropoff_longitude'].values[0]

# Calculate the Haversine distance for the costliest ride
haversine_distance_costliest_ride = haversine(pickup_latitude, pickup_longitude, dropoff_latitude, dropoff_longitude)

print("Haversine distance for the costliest ride:", haversine_distance_costliest_ride, "kilometers")

Haversine distance for the costliest ride: 0.0007899213191009994 kilometers


#### Do you sense something fishy? Try to analyze, and give your expert opinion in Jupyter Notebook.

A Haversine distance of 0.0007899213191009994 kilometers for the costliest ride is exceptionally small, which suggests that there might be an issue or anomaly in the data. It's highly unusual for a ride to have such a short distance while having the highest fare. There could be several reasons for this discrepancy, and further analysis is needed to understand what might have caused this.

Here are some possible explanations:

1. Data Entry Errors: Check if there are any data entry errors or inconsistencies in the latitude and longitude coordinates for this ride. It's possible that incorrect or default values were recorded for the pickup and dropoff locations.

2. Outliers: Examine the 'fare_amount' column to see if there are any outliers or extreme values. Outliers in fare amount can significantly impact the interpretation of this data point.

3. Data Quality: Review the data quality and source to ensure that there were no issues with data collection or preprocessing that could have led to this anomaly.

#### Q13. How many rides were recorded in the year 2014?


In [18]:
df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])

df_cleaned['pickup_year'] = df_cleaned['pickup_datetime'].dt.year

rides_in_2014 = len(df_cleaned[df_cleaned['pickup_year'] == 2014])

print("Number of rides recorded in the year 2014:", rides_in_2014)


Number of rides recorded in the year 2014: 29968


<ipython-input-18-de6c5f767fb0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])
<ipython-input-18-de6c5f767fb0>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['pickup_year'] = df_cleaned['pickup_datetime'].dt.year


#### Q14. How many rides were recorded in the first quarter of 2014?


In [19]:
df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])

df_cleaned['pickup_year_quarter'] = df_cleaned['pickup_datetime'].dt.to_period('Q')

rides_in_Q1_2014 = len(df_cleaned[df_cleaned['pickup_year_quarter'] == '2014Q1'])

print("Number of rides recorded in the first quarter of 2014:", rides_in_Q1_2014)


Number of rides recorded in the first quarter of 2014: 7687


<ipython-input-19-e69275d9b293>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])
C:\Users\Roshan Bhatia\anaconda3\lib\site-packages\pandas\core\arrays\datetimes.py:1101: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  warnings.warn(
<ipython-input-19-e69275d9b293>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['pickup_year_quarter'] = df_cleaned['pickup_datetime'].dt.to_pe

#### Q15. On which day of the week in September 2010, maximum rides were recorded ?

In [20]:
df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])

rides_in_september_2010 = df_cleaned[(df_cleaned['pickup_datetime'].dt.year == 2010) & (df_cleaned['pickup_datetime'].dt.month == 9)]

day_of_week_counts = rides_in_september_2010['pickup_datetime'].dt.day_name().value_counts()

max_rides_day = day_of_week_counts.idxmax()
max_rides_count = day_of_week_counts.max()

print("Day of the week in September 2010 with the maximum number of rides:", max_rides_day)


Day of the week in September 2010 with the maximum number of rides: Thursday


<ipython-input-20-d692edd054f2>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['pickup_datetime'] = pd.to_datetime(df_cleaned['pickup_datetime'])


#### Q16. Apply a Machine Learning Algorithm to predict the fare amount given following input features:passenger_count, distance and ride_week_day.

#### Perform a 70-30 split of data.Which algorithm gives the least adjusted R square value?

In [34]:
df_cleaned['ride_week_day'] = pd.to_datetime(df_cleaned['pickup_datetime']).dt.day_name()

<ipython-input-34-c327c8be4a0b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['ride_week_day'] = pd.to_datetime(df_cleaned['pickup_datetime']).dt.day_name()


In [45]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score, mean_squared_error

In [46]:
df_encoded = pd.get_dummies(df_cleaned, columns=['ride_week_day'])

X = df_encoded[['passenger_count', 'haversine_distance', 'ride_week_day_Friday', 'ride_week_day_Monday', 'ride_week_day_Saturday', 'ride_week_day_Sunday', 'ride_week_day_Thursday', 'ride_week_day_Tuesday']]
y = df_encoded['fare_amount']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree Regressor': DecisionTreeRegressor(),
    'Random Forest Regressor': RandomForestRegressor(),
    'K-Nearest Neighbors Regressor': KNeighborsRegressor()
}

In [47]:
adjusted_r2_values = {}

for model_name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    n = len(X_test)
    p = X_test.shape[1]
    r2 = r2_score(y_test, y_pred)
    
    adjusted_r2 = 1 - (1 - r2) * ((n - 1) / (n - p - 1))
    adjusted_r2_values[model_name] = adjusted_r2

In [48]:
min_adjusted_r2_model = min(adjusted_r2_values, key=adjusted_r2_values.get)
min_adjusted_r2_value = adjusted_r2_values[min_adjusted_r2_model]

print("Adjusted R-squared values:")
for model_name, adjusted_r2 in adjusted_r2_values.items():
    print(f"{model_name}: {adjusted_r2:.4f}")

print(f"The model with the least adjusted R-squared value is {min_adjusted_r2_model} with adjusted R-squared = {min_adjusted_r2_value:.4f}")

Adjusted R-squared values:
Linear Regression: 0.0006
Decision Tree Regressor: 0.4746
Random Forest Regressor: 0.6259
K-Nearest Neighbors Regressor: 0.6340
The model with the least adjusted R-squared value is Linear Regression with adjusted R-squared = 0.0006
